### PR TITLE
SIP-72: dedented triple-quoted string literals 

### PR DESCRIPTION
**Pull request by lihaoyi**
Implements https://github.com/scala/improvement-proposals/pull/112

- The initial Lexing/Scanning is lenient, only looking for the opening `''''*` and equivalent closing delimiter. This matches how we can expect this to be implemented in other tools that have more restricted lexing frameworks (IntelliJ w/ JFlex, VSCode w/ TextMate Grammars, NeoVim w/ TreeSitter)

- All other validation (opening delimiter must be followed by newline, closing delimiter must be preceded by whitespace only) and de-denting is left to the parsing phase, which is the only time we have a complete string "literal" when an interpolator is present, and thus are able to look at the trailing delimiter's preceding indent whitespace and trim it from all earlier `STRINGLIT`/`STRINGPART` tokens

- `def interpolatedString` needed to be refactored to support dedenting: rather than constructing the trees immediately, we first assemble all the strings parts, then use the last string part to compute the dedent that we apply to all other parts, and only then do we construct the trees

- Covered by `neg/` tests and `run/` tests for all the major features and edge cases I could think of: 
    - All indentation removed
    - Some indentation preserved
    - Empty strings
    - Single-line strings
    - Blank lines in the string
    - Leading and trailing blank lines
    - Varying indentation
    - Extensible delimiters with 4 and 5 quotes
    - Funky operator and unicode characters in the string
    - Tab-based indentation
    - Interpolation with `s` and `f`
    - Single- and Multi-line pattern matching with and without interpolation
    - In larger expressions: lists, infix operators, etc.
    - As singleton-type ascriptions and singleton-type parameters
    - As literals passed to `@compileTimeOnly`

- I haven't managed to reliably run tests for some reason, I think I'm bumping into https://contributors.scala-lang.org/t/current-testcompilation/7256. But I tested it manually by copy-pasting the run/neg test files into the `bin/scala` REPL and compared the output manually with the `.check` files on disk, and the output is identical

### CODE REVIEW

**Comment by odersky** [comment]
That was fast!

**Comment by lihaoyi** [comment]
Not ready to review yet! Still need a bit more vibing haha

**Comment by Gedochao** [comment]
> Not ready to review yet! Still need a bit more vibing haha

Ah right, I'll convert it to draft then

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 1]
```diff
@@ -1349,6 +1349,93 @@ object Parsers {
...
+                             isFirstPart: Boolean,
+                             isLastPart: Boolean): String = {
+
+      if (closingIndent == "") str
+      else {
+        // Check for mixed tabs and spaces in closing indent
+
+        val hasTabs = closingIndent.contains('\t')
+        val hasSpaces = closingIndent.contains(' ')
+        if (hasTabs && hasSpaces) {
```
Comment (He-Pin) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: Should be able to detect this in one loop

**Comment by lihaoyi** [comment]
I think this is in a pretty good state right now. There's some funniness around the test infrastructure I haven't figured out, and obviously it would have to wait for https://github.com/scala/improvement-proposals/pull/112, but all the tests pass when run manually.

Despite attempting to vibe code the whole thing, claude could only get about 80% of the way there. The last 20% of feature-work/debugging/cleanup I ended up doing by hand. 

Could certainly use more cleanup, but this is probably the natural stopping point until the SIP goes into the next phase

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 2]
```diff
@@ -1349,6 +1349,93 @@ object Parsers {
       else
         literal(inTypeOrSingleton = true)
 
+    /** Dedent a string literal by removing common leading whitespace.
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: For new code in the compiler we use indentation syntax and new conditional `if / then / else` syntax. The old Java conditional syntax is already disabled under `-language.future`.

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 3]
```diff
@@ -1447,40 +1552,109 @@ object Parsers {
...
-            })
 
-      var offsetCorrection = if isTripleQuoted then 3 else 1
-      while (in.token == STRINGPART)
-        nextSegment(in.offset + offsetCorrection)
+      // Collect all string parts and their offsets
+      val stringParts = new ListBuffer[(String, Offset)]
+      val interpolatedExprs = new ListBuffer[Tree]
+
+      var offsetCorrection = if (isDedented) 3 else if (isTripleQuoted) 3 else 1
```
Comment (lihaoyi) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: This bit is super sketchy, I'm sure there's a better way

**Comment by lihaoyi** [comment]
Marking this as ready to review since the SIP is has been voted into experimental phase

**Comment by lihaoyi** [comment]
@odersky @sjrd looks like most of the tests are green, this should be ready for review. I couldn't find any error message in the failing job's logs, might need help from someone more familiar with the CI setup to help take a look

**Comment by lihaoyi** [comment]
The last failure turned out to be due to error message positioning issues, fixed that and it's all green now

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 4]
```diff
@@ -1447,40 +1536,109 @@ object Parsers {
         in.charOffset + 1 < in.buf.length &&
         in.buf(in.charOffset) == '"' &&
         in.buf(in.charOffset + 1) == '"'
+
+      val isDedented =
+        in.charOffset + 2 < in.buf.length &&
+        in.buf(in.charOffset - 1) == '\'' &&
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: Why do we look at `offset - 1` here? And we should be able to reuse the logic of `isDedentedStringLiteral`?

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 5]
```diff
@@ -1349,6 +1349,78 @@ object Parsers {
...
         literal(inTypeOrSingleton = true)
 
+    /** Dedent a string literal by removing common leading whitespace.
+     *  The amount of whitespace to remove is determined by the indentation
+     *  of the last line (which should contain only whitespace before the
+     *  closing delimiter).
+     *
+     *  @param str The string content to dedent
+     *  @param offset The source offset where the string literal begins
+     *  @return The dedented string, or str if errors were reported
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: Given the importance of other parameters, it would be better to explain them as well, maybe with an example.

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 6]
```diff
@@ -1447,40 +1536,109 @@ object Parsers {
...
+      if (finalLiteral) { // Add the final literal if present
+        val (dedentedStr, offset) = dedentedParts.last
+        segmentBuf += atSpan(offset, offset, offset + dedentedStr.length) { Literal(Constant(dedentedStr)) }
+      }
 
       InterpolatedString(interpolator, segmentBuf.toList)
     }
 
+    /** Extract the closing indentation from the last line of a string */
+    private def extractClosingIndent(str: String, offset: Offset): (String, Boolean) = {
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: I would use an `Option[String]` for the result.

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 7]
```diff
@@ -1349,6 +1349,78 @@ object Parsers {
...
+            em"dedented string literal cannot mix tabs and spaces in indentation",
+            offset
+          )
+          return str
+        }
+
+        val linesAndWithSeps = (str.linesIterator.zip(str.linesWithSeparators)).toSeq
+        var lineOffset = offset
+        // start counting error location offsets only after opening delimiter
+        while(in.buf(lineOffset) == '\'') lineOffset += 1
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: missing a space after `while`

File: tests/run/dedented-string-literals.scala [hunk 8]
```diff
@@ -0,0 +1,288 @@
+// Test runtime behavior of dedented string literals
```
Comment (noti0na1) [inline tests/run/dedented-string-literals.scala]: Can you add the following small tests:
```scala
val x = s'''
${"content with\nnewline"}
more text
'''

val nested = s'''
outer ${'''
  inner
'''}
'''

val nested2 = s'''
outer ${s'''
  inner with $x
'''}
'''
```

**Review by noti0na1 — CHANGES_REQUESTED** [review]

**Comment by lihaoyi** [comment]
@noti0na1 resolved your comments

**Comment by lihaoyi** [comment]
also wired up the REPL incomplete literal detection to work for `'''` strings

**Comment by lihaoyi** [comment]
Bump on this since been a while

**Comment by Gedochao** [comment]
> Bump on this since been a while

No worries, we remember. Work on getting 3.8 to RC1 may be delaying things, sorry about the compiler team being perhaps less a tad responsive these couple days.

**Comment by lihaoyi** [comment]
Just another bump here since it's been two weeks since the last bump, I would like to try and get this into 3.8.0 if possible

**Comment by Gedochao** [comment]
> I would like to try and get this into 3.8.0 if possible

I mean, the SIP is still in progress, 3.8.0-RC1 is already late (the artifacts should become available later today, if all goes well) and this is a rather hefty review (which we haven't jumped on immediately, as finalising 3.8.0-RC1 has devoured most of the team's resources).
So it's almost certainly something that'll wait until 3.10.
That being said, we'll try to get back to the review as soon as possible.

**Comment by lihaoyi** [comment]
The SIP is in progress because it's waiting for an implementation, which is this PR. So in other words the SIP is currently blocked on this review to make progress

Fine if it's impossible to get it in to 3.8.0, just want to make sure my expectations are set appropriately. If you say "a tad less responsive", don't be surprised if I keep pinging you every so often. If you say "come back in a few months", then I'll come back in a few months and stop bothering you haha

**Comment by Gedochao** [comment]
> The SIP is in progress because it's waiting for an implementation, which is this PR. So in other words the SIP is currently blocked on this review to make progress
> 
> Fine if it's impossible to get it in to 3.8.0, just want to make sure my expectations are set appropriately. If you say "a tad less responsive", don't be surprised if I keep pinging you every so often. If you say "come back in a few months", then I'll come back in a few months and stop bothering you haha

By all means, do ping us if we don't come back 😅 

Additional note: 3.10 sounds very distant, but in reality, since 3.9.0 is planned as the new LTS, 3.10.0 will likely follow shortly after, with 3.9.x patches being reserved for backports.
So still distant, just not extremely so.

**Comment by lihaoyi** [comment]
If this goes in experimentslly in 3.10 it won't be usable by the community until non-experimentally in 3.11, which would be in 2027 i assume? That's maybe a year later than going in in 3.8 and becoming usable in 3.9, which is my motivation for trying to get it in. Delaying the SIP a year is a long time for something that's already approved and marked for implementation, and in fact has an implementation ready to go. So you can understand why I'd like to get it in sooner if at all possible

**Comment by He-Pin** [comment]
If this can get merged in 3.8.0 and marked as experimental , then it will always free to change.

**Comment by Gedochao** [comment]
> If this goes in experimentslly in 3.10 it won't be usable by the community until non-experimentally in 3.11, which would be in 2027 i assume?

Yes.

> That's maybe a year later than going in in 3.8 and becoming usable in 3.9, which is my motivation for trying to get it in.

The current intention is to not promote any experimental features to stable or preview between 3.8 and 3.9 (as part of the new LTS feature freeze). Only preview features could get stabilised.

> So you can understand why I'd like to get it in sooner if at all possible

I am aware many features will get delayed as a result of the feature freeze implemented for the new 3.9 LTS' sake. 
We knew this.
That is the cost of making an LTS version to users in a stable state.

Scala 3.8.0-RC1 is long overdue, its release is being prepared as we speak and we will not be delaying it any further. There is always that one more thing that we could potentially get if we wait another couple of days. One has to draw the line at some point. That point is now. Has been for a bit, in fact, as [the 3.8 branch has been cut some time ago](https://github.com/scala/scala3/tree/release-3.8.0), with select changes being back ported there.

Even though we will likely have a 3.8.0-RC2 (if not 3.8.0-RC3, given just how big and complex 3.8 came out to be), subsequent RCs are reserved strictly for fixes and improvements, not new features.

And so, it currently seems unlikely we'd include this PR before 3.10.  
Thank you for your understanding.

**Comment by WojciechMazur** [comment]
We're in the future freeze for the last 2 weeks. It needs to wait for 3.10, be accepted by SIP, made preview in 3.11 and maybe stable in 3.12.

**Comment by odersky** [comment]
It is accepted for implementation so we are OK as far the SIP is concerned. But the fact that we have been in feature freeze for two weeks is true. So the timing does not work out, I am afraid.

**Comment by lihaoyi** [comment]
Got it. If it can't go into 3.8, could it go into 3.9 as experimental? @Gedochao said the restriction for 3.9 is no new stable features, but are experimental features also restricted as @WojciechMazur seems to suggest?

**Comment by odersky** [comment]
Yes, we have a policy of no new experimental features in 3.9. But I don't expect the 3.9 period to last long anyways. Essentially we just want to get to a stable state that we can support long term with backports. Once we are there the head will move to 3.10.

**Comment by odersky** [comment]
@lihaoyi I think it's easiest if I work on the PR a bit to make it more conformant to our internal conventions. I just invited you to https://github.com/dotty-staging/dotty/. Can you push your PR there? Then several people can collaborate on it.

**Comment by lihaoyi** [comment]
Got it! Will wait for 3.10 then. I pushed this PR to https://github.com/dotty-staging/dotty/commit/dedented-strings as you mentioned

**Comment by lihaoyi** [comment]
This PR also has `Allow edits and access to secrets by maintainers` enabled, so feel free to push changes directly here if you would like

**Comment by Gedochao** [comment]
Note: we are entering 3.10, so this might be worth a rebase and a fresh look.

**Comment by lihaoyi** [comment]
@Gedochao @odersky  this is rebased and should be ready for review

**Comment by Gedochao** [comment]
Also pinging @noti0na1 and @sjrd for the review.

**Review by odersky — CHANGES_REQUESTED** [review]
This can now move forward since we are in the 3.10 cycle. But the change set is frighteningly large for a simple review. I propose we push the branch to dotty-staging, then we can collaborate on it together. @lihaoyi you have already write access to this repo, can you move the PR there?

**Comment by lihaoyi** [comment]
@odersky it has been pushed

**Comment by lihaoyi** [comment]
https://github.com/dotty-staging/dotty/commit/dedented-strings

**Comment by lihaoyi** [comment]
@odersky you can also just `git push git@github.com:lihaoyi/scala3.git head:dedented-strings` which will update this PR directly, as maintainers have access to update this PR branch

**Comment by odersky** [comment]
@lihaoyi I ended up re-implementing it by hand, but took the tests from this PR.

There is one thing which is different: This PR allows a quadruple '''' inside a triple quote enclosed string literal. There was a run test for this behavior, which I moved to a neg test in my PR. 

SIP 72 does not address this question explicitly, but since it refers to C# and Swift I assume it implies that this should not be allowed. If we do allow it, then we get surprising behavior for something like

  '''''
    ...
  ''''''    // accidentally wrote one ' too many

this would happily continue to parse and then include everything to the next `'''''` or give an "unclosed string literal". 

There's also a small thing which should be mentioned in the SIP: We used to accept `$"` escapes in interpolated string literals. We now accept either `$"` or `$'` escapes depending on what is the delimiter. I think this makes sense but did not see it mentioned in the spec.

**Comment by lihaoyi** [comment]
> @lihaoyi I ended up re-implementing it by hand, but took the tests from this PR.
> 
> There is one thing which is different: This PR allows a quadruple '''' inside a triple quote enclosed string literal. There was a run test for this behavior, which I moved to a neg test in my PR.
> 
> SIP 72 does not address this question explicitly, but since it refers to C# and Swift I assume it implies that this should not be allowed. If we do allow it, then we get surprising behavior for something like
> 
> ''''' ... '''''' // accidentally wrote one ' too many
> 
> this would happily continue to parse and then include everything to the next `'''''` or give an "unclosed string literal".

Yes I think this probably should not be allowed. Moving the test over sounds great

> There's also a small thing which should be mentioned in the SIP: We used to accept `$"` escapes in interpolated string literals. We now accept either `$"` or `$'` escapes depending on what is the delimiter. I think this makes sense but did not see it mentioned in the spec.

Sounds good to me

**Comment by lihaoyi** [comment]
@odersky maybe you could leave a comment describing at a high level how your implementation differs from the one here? For this one I tried to re-use as much of the existing triple-quoted-string logic and interpolation logic as possible, which is perhaps why it looks a bit scattered since it ends up intertwined with the existing code in two places with some existing code split apart to allow re-use. At a glance your implementation looks a lot more self contained, but it's not obvious to me whether it achieves the same code-sharing that this PR does or whether it does without such code-sharing

**Comment by odersky** [comment]
@lihaoyi I tried to summarize differences at the start of #26332. It also tries to re-use a lot of existing code. At least in the Scanner this is apparent since it added several times fewer new lines.

**Comment by odersky** [comment]
Superseded by #26332
